### DISCUSSION TITLE
Prisma migrate wrongly creates a STRING type for PostgreSQL database

### DISCUSSION THREAD

**Discussion by mg-maia** [prisma/prisma#19650] [Q&A] [answered]
Hi,
I have a Cockroach database maintained by Prisma. I want to switch to PostgreSQL (I know all my data will be lost and I'm ok with that).
My database will be still maintained at Cockroachlabs but I need another instance to be run on a "pure" PostgreSQL.
According to Cockroachlabs documentation, their SQL subset is compatible with PostgreSQL so I should be fine to do it this way.

When I set `provider="postgresql"` in my prisma schema file and run `npx prisma migrate dev --create-only` against my Cockroach database, the SQL migration file contains statements with `STRING` type which are specific to Cockroach.

Here are relevant parts of my schema file and resulting SQL file:

#### `schema.prisma`
```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider          = "postgresql"
  url               = env("DATABASE_URL")
  shadowDatabaseUrl = env("SHADOW_DATABASE_URL")
}

model User {
  id            String    @id @default(cuid())
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map("email_verified")

  @@map("user")
}
```

#### `migration.sql`
```sql
-- CreateTable
CREATE TABLE "user" (
    "id" STRING NOT NULL,
    "name" STRING,
    "email" STRING,
    "email_verified" TIMESTAMP(3),

    CONSTRAINT "user_pkey" PRIMARY KEY ("id")
);
```

### Error
When I run the sql file against Cockroach DB it works fine, but when I try to run the same migration on a "pure" PostgreSQL database, I get this:

```
Database error code: 42704

Database error:
ERROR: type "string" does not exist
```

...which is obvious because `STRING` type does not exist in PostgreSQL. Prisma should generate a `TEXT` type, as stated in the [documentation](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#default-type-mappings).

Why then Prisma creates Cockroach-like statements if the provider is set to `postgresql`?

Does it have anything to do with my previous cockroach instance? I'm using a completely new database on the cockroach cluster and have removed the entire `prisma/` folder. Does Prisma use some kind of a cache?
Or maybe Prisma "figures out" the provider by looking at the database I'm running the migration against? But why then having a `provider` field in the schema?

**Comment by ludralph** [comment]
Hi @mg-maia 👋 

Prisma does not use any caching mechanism. Can you manually edit the migration file after you have initiated this command `npx prisma migrate dev --create-only`? Edit the `STRING` to `TEXT` and try applying the migration again.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by mg-maia** [reply]
Hi @ludralph ,

I could do that but it wouldn't solve the problem in a long term. That would basically mean, every time I create a migration I'd have to remember about going to that SQL file and replace all `STRING`s with `TEXT`s. It also means I'd be forced to always use the `--create-only` flag.

Are you able to explain why Prisma still tries to force the Cockroach specific language? It looks to me like a bug in Prisma itself, unless you have a better explanation.

**Comment by janpio** [comment]
> Or maybe Prisma "figures out" the provider by looking at the database I'm running the migration against? But why then having a provider field in the schema?

Because we believed "According to Cockroachlabs documentation, their SQL subset is compatible with PostgreSQL so I should be fine to do it this way." for a while, before we learned that not to be true at all in our context and we introduced an additional, specific provider for CockroachDB.

I suggest you test against a local CockroachDB instance as well if you want to run production on CRDB. Everything else will be a lot of workarounds and probably pain.

**Reply by mg-maia** [reply, marked as answer]
Hi @janpio,

I thought the same when I read it. But now I think what they meant is that the CRDB is a _superset_ of PostgreSQL.
Though, looks like the problem I've described, remains.

Anyway, I guess I'll go with a local CRDB cluster. Thanks for that suggestion!
